configs/module_test.go

```diff
@@ -203,3 +203,92 @@ func TestModule_required_provider_overrides(t *testing.T) {
 		)
 	}
 }
+
+// Resources without explicit provider configuration are assigned a provider
+// implied based on the resource type. For example, this resource:
+//
+//  resource foo_instance "test" { }
+//
+// is assigned a provider with type "foo".
+//
+// To find the correct provider, we first look in the module's provider
+// requirements map for a local name matching the resource type, and fall back
+// to a default provider if none is found. This applies to both managed and
+// data resources.
+func TestModule_implied_provider(t *testing.T) {
+	mod, diags := testModuleFromDir("testdata/valid-modules/implied-providers")
+	if diags.HasErrors() {
+		t.Fatal(diags.Error())
+	}
+
+	// The two providers used in the config resources
+	foo := addrs.NewProvider("registry.acme.corp", "acme", "foo")
+	bar := addrs.NewDefaultProvider("bar")
+
+	// Verify that the registry.acme.corp/acme/foo provider is defined in the
```

Comment:
I'm slightly confused by the intent of this test after reading the command and looking at the test module; I tend to think of "implied provider" as meaning there is no matching provider configuration. 

We have three scenarios, spelling them out helped *me* understand my confusion:
* explicitly configured `provider` attr in the resource
* implied provider (no matching required_provider) : no `provider` attribute or matching required_providers block 
* implied provider (matching local name): there's a required_providers local name that matches the implied type

We have two variations on "implied": with or without a matching required_providers localname.

If this test is confirming that the *local name* maps properly, it might be clearer for future readers if the local name is not the same as the type name (for example, `myfoo = "registry.acme.corp/acme/foo"`).

Reply:
This is really good feedback, thanks @mildwonkey!

You're right that the test is confirming that the resource's implied provider type name is mapped to the configured provider local name (whew, that's a mouthful). Using a provider with FQN type name which matches the local name is definitely confusing.

[In my latest push, I've added another provider and resources to the test cases](https://github.com/hashicorp/terraform/compare/ecd52dc0f75a326bb8b3e103f1fa87e77fd36da5..03059e4516035eb41b4c7a99474db5381492b67b), along with a few more comments. The provider is configured as:

```tf
whatever = {
  source = "acme/something"
}
```

Therefore it's used as `whatever_resource`.

I decided to keep the `foo` example around for now though, because I think it more accurately reflects common usage of the feature. I'm thinking of it as a sort of future-me documentation, rather than code coverage.

My guess is that if a provider type name doesn't match the configuration's provider local name, implied provider mapping probably won't work. Custom provider resource addresses probably still start with the type name. Users in this situation are probably going to end up needing to use resource-level `provider` attributes.

Does all this make sense to you? If I've misunderstood common usage, the `foo` examples might be misleading and I should remove them.

Reply:
It does make sense, thank you! Earlier in this project I spent a lot of time crafting tests that would isolate the local name mapping specifically, and others that were testing the implied type (with or without matching local name), so I was just confusing myself. Looks great 🎉